File: lib/src/display_manager.dart
import 'dart:ffi' as ffi;
import 'dart:ui';

import 'package:ffi/ffi.dart';
import 'package:nativeapi/src/display_listener.dart';
import 'package:nativeapi/src/event_listener_mixin.dart';
import 'package:nativeapi/src/nativeapi_bindings_generated.dart';

import 'display.dart';
import 'nativeapi_bindings.dart';

class DisplayManager with EventListenerMixin<DisplayListener> {
  DisplayManager._();

  static final DisplayManager instance = DisplayManager._();

  NativeApiBindings get _bindings => nativeApiBindings;

  late final ffi.NativeCallable<DisplayAddedCallbackFunction>
      _addedCallbackCallable;
  late final ffi.NativeCallable<DisplayRemovedCallbackFunction>
      _removedCallbackCallable;

  DisplayManager() {
    _addedCallbackCallable =
        ffi.NativeCallable<DisplayAddedCallbackFunction>.listener(
            _onDisplayAdded);
    _removedCallbackCallable =
        ffi.NativeCallable<DisplayRemovedCallbackFunction>.listener(
            _onDisplayRemoved);

    _bindings.display_manager_on_display_added(
        _addedCallbackCallable.nativeFunction);
    _bindings.display_manager_on_display_removed(
        _removedCallbackCallable.nativeFunction);
  }

  void _onDisplayAdded(NativeDisplay display) {
    print('onDisplayAdded: ${display.dartify()}');
    notifyListeners((l) => l.onDisplayAdded(display.dartify()));
  }

  void _onDisplayRemoved(NativeDisplay display) {
    print('onDisplayRemoved: ${display.dartify()}');
    notifyListeners((l) => l.onDisplayRemoved(display.dartify()));
  }

  /// Get all displays.
  ///
  /// Returns a list of all displays.
  List<Display> getAll() {
    return _bindings.display_manager_get_all().dartify();
  }

  /// Get the primary display.
  ///
  /// Returns the primary display.
  Display getPrimary() {
    return _bindings.display_manager_get_primary().dartify();
  }

  /// Get the cursor position.
  ///
  /// Returns the cursor position in the primary display.
  Offset getCursorPosition() {
    return _bindings.display_manager_get_cursor_position().dartify();
  }

  @override
  void addListener(DisplayListener listener) {
    super.addListener(listener);
    if (hasListeners) _bindings.display_manager_start_listening();
  }

  @override
  void removeListener(DisplayListener listener) {
    super.removeListener(listener);
    if (!hasListeners) _bindings.display_manager_stop_listening();
  }
}

extension on NativeDisplay {
  Display dartify() {
    String id, name;
    try {
      id = this.id.cast<Utf8>().toDartString();
    } catch (e) {
      id = '';
    }
    try {
      name = this.name.cast<Utf8>().toDartString();
    } catch (e) {
      name = '';
    }

    return Display(
      id: id,
      name: name,
      size: Size(width, height),
      scaleFactor: scaleFactor,
    );
  }
}

extension on NativeDisplayList {
  List<Display> dartify() {
    return List.generate(count, (index) => displays[index].dartify());
  }
}

extension on NativePoint {
  Offset dartify() {
    return Offset(x, y);
  }
}

File: src/display_manager.cpp
#include <string.h>
#include <iostream>

#include "display_manager.h"
#include "libnativeapi/include/nativeapi.h"

using namespace nativeapi;

struct NativeDisplay to_native_display(const Display& display) {
  NativeDisplay native_display;
  native_display.id = strdup(display.id.c_str());
  native_display.name = strdup(display.name.c_str());
  native_display.width = display.width;
  native_display.height = display.height;
  native_display.visiblePositionX = display.visiblePositionX;
  native_display.visiblePositionY = display.visiblePositionY;
  native_display.visibleSizeWidth = display.visibleSizeWidth;
  native_display.visibleSizeHeight = display.visibleSizeHeight;
  native_display.scaleFactor = display.scaleFactor;
  return native_display;
}

static DisplayAddedCallback g_display_added_callback = nullptr;
static DisplayRemovedCallback g_display_removed_callback = nullptr;

DisplayManager g_display_manager = DisplayManager();
DisplayEventHandler g_display_event_handler = DisplayEventHandler(
    [](const Display& display) {
      std::cout << "Display added: " << display.id << std::endl;
      if (g_display_added_callback) {
        g_display_added_callback(to_native_display(display));
      }
    },
    [](const Display& display) {
      std::cout << "Display removed: " << display.id << std::endl;
      if (g_display_removed_callback) {
        g_display_removed_callback(to_native_display(display));
      }
    });

FFI_PLUGIN_EXPORT
struct NativeDisplayList display_manager_get_all() {
  auto displays = g_display_manager.GetAll();
  NativeDisplayList native_display_list;
  native_display_list.count = displays.size();
  for (size_t i = 0; i < displays.size(); i++) {
    native_display_list.displays[i] = to_native_display(displays[i]);
  }
  return native_display_list;
}

FFI_PLUGIN_EXPORT
struct NativeDisplay display_manager_get_primary() {
  auto display = g_display_manager.GetPrimary();
  return to_native_display(display);
}

FFI_PLUGIN_EXPORT
struct NativePoint display_manager_get_cursor_position() {
  auto cursorPosition = g_display_manager.GetCursorPosition();
  NativePoint native_point;
  native_point.x = cursorPosition.x;
  native_point.y = cursorPosition.y;
  return native_point;
}

FFI_PLUGIN_EXPORT
void display_manager_start_listening() {
  g_display_manager.RemoveListener(&g_display_event_handler);
  g_display_manager.AddListener(&g_display_event_handler);
}

FFI_PLUGIN_EXPORT
void display_manager_stop_listening() {
  g_display_manager.RemoveListener(&g_display_event_handler);
}

FFI_PLUGIN_EXPORT
void display_manager_on_display_added(DisplayAddedCallback callback) {
  g_display_added_callback = callback;
}

FFI_PLUGIN_EXPORT
void display_manager_on_display_removed(DisplayRemovedCallback callback) {
  g_display_removed_callback = callback;
}
